File: 690.employee-importance.py
#
# @lc app=leetcode id=690 lang=python3
#
# [690] Employee Importance
#
"""
# Employee info
class Employee:
    def __init__(self, id, importance, subordinates):
        # It's the unique id of each node.
        # unique id of this employee
        self.id = id
        # the importance value of this employee
        self.importance = importance
        # the id of direct subordinates
        self.subordinates = subordinates
"""
class Solution:
    def getImportance(self, employees, id):
        """
        :type employees: Employee
        :type id: int
        :rtype: int
        """
        d = dict()
        for node in employees:
            d[node.id] = node

        def Importance(node):
            if len(node.subordinates) == 0:
                return node.importance
            else:
                S = node.importance
                for _id in node.subordinates:
                    S += Importance(d[_id])
                return S
        return Importance(d[id])

        

File: 690.employee-importance.java
/*
 * @lc app=leetcode id=690 lang=java
 *
 * [690] Employee Importance
 */
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    public int getImportance(List<Employee> employees, int id) {
        Map<Integer, Employee> map = new HashMap<>();
        for(Employee employee : employees){
            map.put(employee.id, employee);
        }
        return importance(map, id);
    }

    private int importance(Map<Integer, Employee> map, int id){
        Employee employee = map.get(id);
        int total = employee.importance;

        for(int subordinateId : employee.subordinates){
            total += importance(map, subordinateId);
        }
        return total;
    }
}

